app/demo-loop/diagnostic/StepDiagnostic.i.hh

```diff
@@ -0,0 +1,159 @@
+std::unordered_map<std::string, std::vector<size_type>>
+StepDiagnostic<M>::steps()
+{
+    using BinId = celeritas::ItemId<size_type>;
+
+    // Copy result to host if necessary
+    StepDiagnosticData<Ownership::value, MemSpace::host> data;
+    data = data_;
+
+    // Map particle ID to particle name and store steps per track distribution
+    std::unordered_map<std::string, std::vector<size_type>> result;
+    for (auto particle_id : range(celeritas::ParticleId{particles_->size()}))
+    {
+        auto start = BinId{particle_id.get() * data.max_steps};
+        auto end   = BinId{start.get() + data.max_steps};
+        CELER_ASSERT(end.get() <= data.counts.size());
+        auto counts = data.counts[celeritas::ItemRange<size_type>{start, end}];
+
+        if (std::any_of(counts.begin(), counts.end(), [](size_type x) {
+                return x > 0;
+            }))
+        {
+            result[particles_->id_to_label(particle_id)]
+                = {counts.begin(), counts.end()};
+        }
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+template<MemSpace M>
+CELER_FUNCTION StepLauncher<M>::StepLauncher(const ParamsDataRef&     params,
+                                             const StateDataRef&      states,
+                                             StepDiagnosticDataRef<M> data)
+    : params_(params), states_(states), data_(data)
+{
+    CELER_EXPECT(params_);
+    CELER_EXPECT(states_);
+    CELER_EXPECT(data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Increment step count and tally number of steps for tracks that were killed.
+ *
+ * At this point, active tracks will either be marked "alive" in the sim state
+ * or will have a killing action in their interaction result. Tracks that were
+ * killed in a discrete interaction will still be marked as alive (these are
+ * killed in post-processing), but their action will be killing.
+ */
+template<MemSpace M>
+CELER_FUNCTION void StepLauncher<M>::operator()(celeritas::ThreadId tid) const
+{
+    using BinId = celeritas::ItemId<size_type>;
+
+    celeritas::ParticleTrackView particle(
+        params_.particles, states_.particles, tid);
+    celeritas::SimTrackView sim(states_.sim, tid);
+
+    const auto& interaction = states_.interactions[tid];
+
+    // Increment the step count if this is an active track
+    if (sim.alive() || celeritas::action_killed(interaction.action))
+    {
+        ++data_.steps[tid];
+    }
+
+    // If the track was killed, tally the number of steps and reset the track's
+    // step counter.
+    if (celeritas::action_killed(interaction.action))
+    {
+        CELER_ASSERT(data_.steps[tid] > 0);
+        if (data_.steps[tid] <= data_.max_steps)
+        {
+            auto index = particle.particle_id().get() * data_.max_steps
+                         + data_.steps[tid] - 1;
```

Comment:
This indexing operation happens in a couple of places -- ideally we'd move to an `ndarray`-type class that allows us to index into these, but for now maybe a small (anonymous?) helper function?